### PR TITLE
Add support of Facebook Infer static analyzer

### PR DESCRIPTION
**Pull request by begarco**
## Proposed changes

This pull request only adds support of the Infer static analyzer. It is able to import the JSON report containing warnings about potential source of bugs.

Supported version of Infer: `0.17.0`

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update

## Checklist
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove the feature works
- [x] I have added necessary documentation (if appropriate)

---
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1868)

### CODE REVIEW

**Comment by guwirth** [comment]
Hello @begarcom,

Thx for providing this. I will review it ASAP. 

Regards,

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    InferIssue[] inferIssues;
+
+    try (JsonReader reader = new JsonReader(new FileReader(report))) {
+      Gson gson = new Gson();
+      inferIssues = gson.fromJson(reader,InferIssue[].class);
+    }
+
+    for(InferIssue issue : inferIssues) {
+      LOG.debug("Read: {}", issue.toString());
+      if(sensor.getInputFileIfInProject(issue.getFile()) != null) {
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java:61]: `getInputFileIfInProject` is an expensive call and `saveUniqueViolation `is doing the same already internally. I would remove it here and validate only if a file is set. There are also project issues possible **without** file, but I think Infer is not supporting this?

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/CxxInferSensor.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @Override
+  protected void processReport(File report) {
+    InferParser parser = new InferParser(this);
+    try {
+      parser.processReport(report);
+      LOG.info("Added report '{}' (parsed by: {})", report, parser);
+    } catch (IOException e) {
+      LOG.error("Report {} cannot be found", report);
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/CxxInferSensor.java:75]: has to be handled more accurate in future with #1866

**Comment by guwirth** [comment]
@begarco to document this feature it would be helpful:
- link to infer tool?
- with which version of the tool did you the tests?
- how to call / use the tool to generate the report?
- link to report file format?

**Comment by begarco** [comment]
@guwirth I push change asap

For other questions:
- Infer official web site: https://fbinfer.com/
- Supported version: `0.17.0` (the latest)
- Usage: `infer run --report-force-relative-path --linters --keep-going -- make` (you can replace make by your own build command)
- Report file format: `JSON` as defined [here](https://github.com/facebook/infer/blob/348a3927490b15c75ecd5bc6ff25bf56f64e6675/infer/src/integration/JsonReports.ml#L202)

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java (hunk 1 above)
Comment (begarco) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java]: Fixed in https://github.com/SonarOpenCommunity/sonar-cxx/pull/1868/commits/ed88154c3b636ac99aa689ef76b16d8b5733da97

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/CxxInferSensor.java (hunk 2 above)
Comment (begarco) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/CxxInferSensor.java]: Fixed in https://github.com/SonarOpenCommunity/sonar-cxx/pull/1868/commits/ed88154c3b636ac99aa689ef76b16d8b5733da97

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    for(InferIssue issue : inferIssues) {
+      LOG.debug("Read: {}", issue.toString());
+      if(issue.getFile() != null) {
+        CxxReportIssue cxxReportIssue = new CxxReportIssue(
+                issue.getBugType(), issue.getFile(),
+                String.valueOf(issue.getLine()), issue.getQualifier());
+        sensor.saveUniqueViolation(cxxReportIssue);
+      } else {
+        LOG.debug("Cannot find the file '{}', skipping violations", issue.getFile());
```
Comment (guwirth) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/infer/InferParser.java:67]: @begarco think this message is wrong? It’s more Invalid infer issue/message... file as parameter makes also no sense, is null. Would add issue.toString.

**Comment by guwirth** [comment]
@begarco thanks a lot for providing this PR

**Pull request merged**